src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -163,37 +166,86 @@ private void readObject(ObjectInputStream pIn) throws IOException, ClassNotFound
   private static final double ONE_HALF_LOG = Math.log(0.5);
   private Random randomGenerator = new Random();
 
-  private final Comparator<? super T> comparator;
+  private final @Nullable Comparator<? super T> comparator;
   private Node<T> head = createHead();
   private Node<T> tail = head;
   private int size = 0;
 
-  @SuppressWarnings("unchecked")
-  public SkipList(@Nullable Comparator<? super T> pComparator) {
-    if (pComparator == null) {
-      comparator = (Comparator<? super T>) Comparator.naturalOrder();
-    } else {
-      comparator = pComparator;
-    }
+  private SkipList(Comparator<? super T> pComparator) {
+    Preconditions.checkNotNull(pComparator);
+    comparator = pComparator;
   }
 
-  public SkipList(Collection<? extends T> pCollection) {
+  private SkipList(Iterable<? extends T> pCollection) {
     this();
-    boolean changed = addAll(pCollection);
-    assert changed;
+    addAll0(pCollection);
   }
 
-  public SkipList(SortedSet<T> pSortedSet) {
-    this(pSortedSet.comparator());
-    boolean changed = addAll(pSortedSet);
-    assert changed;
+  private SkipList(Iterable<? extends T> pCollection, Comparator<? super T> pComparator) {
+    this(pComparator);
+    addAll0(pCollection);
   }
 
-  @SuppressWarnings("unchecked")
-  public SkipList() {
-    comparator = (Comparator<? super T>) Comparator.naturalOrder();
+  private SkipList() {
+    comparator = null;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1) {
+    SkipList<T> list = new SkipList<>();
+    list.add(p1);
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2));
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2, T p3) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2, p3));
+    return list;
   }
 
+  @SafeVarargs
+  public static <T extends Comparable<T>> SkipList<T> of(T p1, T p2, T p3, T... pOtherElements) {
+    SkipList<T> list = new SkipList<>();
+    list.addAll(Lists.newArrayList(p1, p2, p3));
+    for (T v : pOtherElements) {
+      list.add(v);
+    }
+    return list;
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> create() {
+    return new SkipList<>();
+  }
+
+  public static <T> SkipList<T> create(Comparator<? super T> pComparator) {
+    return new SkipList<>(pComparator);
+  }
+
+  public static <T> SkipList<T> create(
+      Iterable<? extends T> pCollection, Comparator<? super T> pComparator) {
+    return new SkipList<>(pCollection, pComparator);
+  }
+
+  public static <T extends Comparable<T>> SkipList<T> createWithNaturalOrder(
```

Comment:
How about `copyOfWithNaturalOrder`?

Reply:
I had that first, but 'copy' implies that it's copied - and is a collection really copied if the position of its elements could be completely changed?

Reply:
That's maybe a valid point, but on the other hand the `WithNaturalOrder` makes it pretty clear what happens. 

I don't think that `create(Iterable)` really conveys its semantics well when used: Is it really obvious that `SkipList.create(foos)` returns a collection that contains the same as `foos`?

So maybe we find a third alternative?

Reply:
How about `create(Iterable)` to create with natural order,
and `createOfSorted(SortedSet)` or `createWithSameOrder(SortedSet)`?

Reply:
I would agree. Is it clear that `createWithSameOrder()` also copies the content and not only the order?

Reply:
Good point.

I would prefer `copyOfWithSameOrder`, similar to what I used before, but that would be even less consistent with the Guava convention.

Even though it is less than optimal, I think we can also assume that the parameter implies that the elements of the given list will be copied (otherwise, why not use a `Comparator`-parameter)?
But maybe you have some other good idea for the naming.